File: server/internal/server/objects/units/Skills/UniversalSkill.go
package Skills

import (
	"server/internal/server"
	"server/internal/server/combatDSL"
	"server/internal/server/objects"
	"server/pkg/packets"
)

type UniversalSkill struct {
	UnitId             int32
	ClientId           uint64
	Client             *server.Client
	EnemyClient        *server.Client
	GameService        GameSkillService
	GameTerrainService GameTerrainService
	data               Data
	ActionCompiler     *combatDSL.SkillCompiler
	Action             combatDSL.CombatAction
}

func (u *UniversalSkill) Type() packets.SkillType {
	return packets.SkillType_None
}

func (u *UniversalSkill) Name() string {
	return "UniversalSkill"
}

func (u *UniversalSkill) Initiate(UnitId int32, client *server.Client, enemyClient *server.Client, gameService GameSkillService) {
	u.UnitId = UnitId
	u.Client = client
	u.EnemyClient = enemyClient
	u.GameService = gameService
	u.GameTerrainService = gameService.World()
	u.ActionCompiler = combatDSL.NewSkillCompiler(nil)
}

func (u *UniversalSkill) CombatString(str string) {
	Action, err := u.ActionCompiler.CompileText(str)
	if err != nil {
		panic(err)
	}
	u.Action = Action

}
func (u *UniversalSkill) ValidTarget(pos objects.Vector2I) bool {
	return true
}

func (u *UniversalSkill) Use(skillId int32, pos objects.Vector2I) {
	tile := u.GameTerrainService.GetTileAt(pos)
	unit := tile.Unit

	u.Action.Execute(combatDSL.CombatContext{
		Caster: unit,
		Target: unit,
	}, skillId)
}

func (u *UniversalSkill) SendPacket(packet packets.Msg) {
	//TODO implement me
	panic("implement me")
}

File: Client/Objects/Units/Skill/UniversalSkill.cs
using Godot;
using Packets;
using Packets.Util;


public class UniversalSkill : ISkill
{
    public IUnit Unit { get; set; }
    public ICombatAction CombatAction { get; set; }
    public SkillDataJSON Data { get; set; }
    public virtual string Name()
    {
        return "UniversalSkill";
    }

    public Node3D Node { get; set; }
    public virtual void Init(IUnit unit, Node3D node, SkillDataJSON data)
    {
        Node = node;
        Unit = unit;
        Data = data;
    }

    public virtual void CombatString(string str)
    {
        CombatAction = Globals.GM.SkillCompiler.Compile(str);
    }
    
    public virtual SkillType Type()
    {
        return Data.Type;
    }

    public virtual void Use(Vector2I position)
    {
        var skillId = Globals.GM.CurrentGameData.CurrentSkillID;
        CombatAction.Execute(new CombatContext(Unit, GetUnitAt(position)), skillId);
        // we need to send the skillId with the packet
        SendPacket(position, skillId);
    }

    protected IUnit GetUnitAt(Vector2I position)
    {
        return Globals.GM.CurrentGameData.GetTileAt(position).Unit;
    }
    public virtual void SendPacket(Vector2I position, int skillId)
    {
        TrafficManager.Send(PacketUtil.NewSkillPacket(skillId, Name(), Type(), Unit.Data.UnitId, PacketUtil.WrapVec2I(position)));
    }
}